### DISCUSSION TITLE
How to use quirks?

### DISCUSSION THREAD

**Discussion by Lionstiger** [erlef/oidcc#342] [Q&A] [answered]
Hi,
I have been trying to setup a local development environment using phoenix and the `phx_gen_oidcc` tool using a local authentik server for oauth2. Still fairly new to elixir/phoenix so this is likely a very basic question.

mix phx.server fails to start with the error
`TLS :client: In state :certify at ssl_handshake.erl:2136 generated CLIENT ALERT: Fatal - Bad Certificate`
indicating that it requires a non selfsigned certs.
Discussion #276 seems to have the solution.

But i cant seem to find the right place/format to provide the `allow_unsafe_http` quirk.

My current relevant block of runtime.exs (where `phx_gen_oidcc` places the other config items):
```elixir
config :argument, Oidcc,
  issuer: "https://localhost:9443/application/o/elixir-test/",
  client_id: "<client_id>",
  client_secret:   "<client_secret>",
  quirks: %{:allow_unsafe_http => true}
```

How/Where do i have to specify the `allow_unsafe_http` quirk for it to be used correctly?

**Comment by maennchen** [comment, marked as answer]
Currently that value is not read from the config. You'll have to supply it when starting the provider configuration.

However: The `quirk` allows the `http` protocol, not the `https` protocol with an unknown certificate. You'll have to configure the `ssl` options for that.


Have a look at:

* https://hexdocs.pm/oidcc/oidcc_provider_configuration_worker.html#t:opts/0
* https://hexdocs.pm/oidcc/oidcc_http_util.html#t:request_opts/0

You can see the quirks in use in the tests: 

https://github.com/erlef/oidcc/blob/ebae4bd01556f07dae446fcd5ea266258cadbb15/test/oidcc_provider_configuration_SUITE.erl#L119-L125

**Reply by Lionstiger** [reply]
Thank you!
authentik seems to also provide a http connection, so for now I'll just use that one, since its just local dev for now.

For future reference I needed to make the quirk change in `lib/<project-name>/application.ex` :
```elixir
{Oidcc.ProviderConfiguration.Worker,
       %{
         issuer: Application.fetch_env!(:argument, Oidcc)[:issuer],
         name: <project-name>.OpenIdConfigurationProvider,
         provider_configuration_opts: %{
           :quirks => %{:allow_unsafe_http => true}
         }
}},
```
and I changed the block in `config/runtime.exs` to look like this:
```elixir
config :argument, Oidcc,
  issuer: "http://localhost:9000/application/o/elixir-test/",
  client_id: "<client_id>",
  client_secret:   "<client_secret>",
```

Still a little confused by the erlang/elixir differences, but I'm glad its working now.

**Reply by maennchen** [reply]
@Lionstiger Great 😊

You should be able to only use the elixir modules. For the types, erlang modules are referenced. But those are only different in syntax. For example:

`%{allow_unsafe_http: true}` instead of `#{allow_unsafe_http => true}`

If you have specific uncertainties about the difference, I’d be happy to address those im the documentation.

**Reply by Lionstiger** [reply]
I will try the different spelling tomorrow. Still reading up on the syntax differences between elixir and erlang.
Thank you for your help and the offer to add to the documentation.

**Reply by Lionstiger** [reply]
Can confirm that works.
```
provider_configuration_opts: %{
    :quirks => %{allow_unsafe_http: true}
}
```
both `allow_unsafe_http` and `true` are atoms but can be used without the `:` prefix.
Thanks for the tip!

**Reply by maennchen** [reply]
quirks is also an atom 😊

**Reply by Lionstiger** [reply]
yes, but i couldn't remove the : prefix, without causing an error, so i assumed the syntax requires the prefix.

**Reply by maennchen** [reply]
```elixir
provider_configuration_opts: %{
    quirks: %{allow_unsafe_http: true}
}
```

`%{atom: value}` is equivalent to `%{:atom => value}`

**Reply by Lionstiger** [reply]
wow, so you just also have to remove the : after quirks. Thanks again for all the help!
